### PR TITLE
session: fix data race in ParseWithParamsInternal

### LINKED ISSUE #30918: DATA RACE in the SessionVars.InRestrictedSQL
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

make race

```
WARNING: DATA RACE
Read at 0x00c001752e3b by goroutine 70:
  github.com/pingcap/tidb/distsql.(*selectResult).fetchResp.func1()
      /go/tidb/distsql/select_result.go:164 +0x18f
  github.com/pingcap/tidb/distsql.(*selectResult).fetchResp()
      /go/tidb/distsql/select_result.go:251 +0xcd7
  github.com/pingcap/tidb/distsql.(*selectResult).Next()
      /go/tidb/distsql/select_result.go:257 +0xfa
  github.com/pingcap/tidb/executor.(*tableResultHandler).nextChunk()
      /go/tidb/executor/table_reader.go:433 +0x1a8
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).Next()
      /go/tidb/executor/table_reader.go:234 +0x304
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:286 +0x58f
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchProbeSideChunks()
      /go/tidb/executor/join.go:214 +0x2d6
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndProbeHashTable.func1()
      /go/tidb/executor/join.go:338 +0xc8
  github.com/pingcap/tidb/util.WithRecovery()
      /go/tidb/util/misc.go:100 +0x77
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndProbeHashTable·dwrap·82()
      /go/tidb/executor/join.go:336 +0x47
Previous write at 0x00c001752e3b by goroutine 78:
  github.com/pingcap/tidb/session.(*session).ParseWithParamsInternal()
      /go/tidb/session/session.go:1532 +0x10e
  github.com/pingcap/tidb/executor.getRowCountAllTable()
      /go/tidb/executor/infoschema_reader.go:193 +0xad
  github.com/pingcap/tidb/executor.(*statsCache).get()
      /go/tidb/executor/infoschema_reader.go:299 +0x1d1
  github.com/pingcap/tidb/executor.(*memtableRetriever).setDataFromTables()
      /go/tidb/executor/infoschema_reader.go:520 +0xb0
  github.com/pingcap/tidb/executor.(*memtableRetriever).retrieve()
      /go/tidb/executor/infoschema_reader.go:107 +0x824
  github.com/pingcap/tidb/executor.(*MemTableReaderExec).Next()
      /go/tidb/executor/memtable_reader.go:119 +0x407
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:286 +0x58f
  github.com/pingcap/tidb/executor.(*SelectionExec).Next()
      /go/tidb/executor/executor.go:1329 +0x1d8
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:286 +0x58f
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows()
      /go/tidb/executor/join.go:282 +0x39d
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2()
      /go/tidb/executor/join.go:731 +0xf6
  github.com/pingcap/tidb/util.WithRecovery()
      /go/tidb/util/misc.go:100 +0x77
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable·dwrap·90()
      /go/tidb/executor/join.go:728 +0x47
Goroutine 70 (running) created at:
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndProbeHashTable()
      /go/tidb/executor/join.go:336 +0x288
  github.com/pingcap/tidb/executor.(*HashJoinExec).Next()
      /go/tidb/executor/join.go:689 +0x1225
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:286 +0x58f
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchProbeSideChunks()
      /go/tidb/executor/join.go:214 +0x2d6
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndProbeHashTable.func1()
      /go/tidb/executor/join.go:338 +0xc8
  github.com/pingcap/tidb/util.WithRecovery()
      /go/tidb/util/misc.go:100 +0x77
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndProbeHashTable·dwrap·82()
      /go/tidb/executor/join.go:336 +0x47
Goroutine 78 (running) created at:
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable()
      /go/tidb/executor/join.go:728 +0x3d7
  github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func1()
      /go/tidb/executor/join.go:687 +0xc8
  github.com/pingcap/tidb/util.WithRecovery()
      /go/tidb/util/misc.go:100 +0x77
  github.com/pingcap/tidb/executor.(*HashJoinExec).Next·dwrap·88()
      /go/tidb/executor/join.go:685 +0x47 
```

https://prow.hawkingrei.com/view/gs/prowpingcap/logs/tidb_race/1473216687745536000

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #30918 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
fix data race in `ParseWithParamsInternal`
```

### CODE REVIEW

**Comment by Defined2014** [comment]
/cc @wjhuang2016 @xhebox @tiancaiamao @hawkingrei
PTAL

File: session/session.go [hunk 1]
```diff
@@ -1528,14 +1528,16 @@ func (s *session) ParseWithParams(ctx context.Context, sql string, args ...inter
...
-	defer func() {
-		s.sessionVars.InRestrictedSQL = origin
-	}()
-	return s.ParseWithParams(ctx, sql, args...)
+	tmp, err := s.sysSessionPool().Get()
+	if err != nil {
+		return nil, err
+	}
+	defer s.sysSessionPool().Put(tmp)
+	se := tmp.(*session)
```
Comment (tiancaiamao) [inline session/session.go:1538]: The design of this function is problematic, the parse and the execute is not using the same session.
If the SQL has some side-effect, for example, a warning message, it's left on the pool's session.

A better solution is a callback style API:

```
func (s *session) runWithInternalSQLExec(fn func(exec sqlexec.RestrictedSQLExec) error) {
    tmp, err := s.sysSessionPool().Get()
	if err != nil {
		return nil, err
	}
	defer s.sysSessionPool().Put(tmp)
    se := tmp.(*session)
    return fn(se)
}
```

The call use it like this:

```
se.runWithInternalSQLExec(func(exec *exec.RestrictedSQLExec) {
    stmt := exec.ParseWithParamsInternal(...)
    exec.Exec(stmt)
    ...
})
```

And you can finally wrap a simpler API like:

```
func (s *session) ExecRestrictedSQL(sql string) {
    s.runWithInternalSQLExec(func(exec sqlexec.RestrictedSQLExec) {
        stmt := exec.ParseWithParamsInternal(...)
        exec.Exec(stmt)
        ...        
    })
}
```

The simplied API is more convenient, but its design is not secure.

Comment (tiancaiamao) [reply session/session.go:1538]: > But we didn't use same session before. And I don't know it's a bug or feature 😭.

Before a previous refactoring, they're in the same session.
The data race is introduced after that refactoring

**Comment by tiancaiamao** [comment]
@hawkingrei  are you also working on this? we should avoid repeated working

**Comment by hawkingrei** [comment]
> @hawkingrei are you also working on this? we should avoid repeated working

No, I just fix the data race problem.

**Comment by hawkingrei** [comment]
> > @hawkingrei are you also working on this? we should avoid repeated working
> 
> No, I am not refactoring. I just fix the data race problem.

I can provide the necessary help.

File: session/session.go (hunk 1 above)
Comment (Defined2014) [reply session/session.go]: Fix the data race fisrt, it may affects 5.4.0.

**Review by wjhuang2016 — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

**Pull request merged**